llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -980,6 +980,27 @@ static Instruction *foldIDivShl(BinaryOperator &I,
       Ret = BinaryOperator::CreateSDiv(X, Y);
   }
 
+  // If X << Y and X << Z does not overflow, then:
+  // (X << Y) / (X << Z) -> (1 << Y) / (1 << Z) -> 1 << Y >> Z
+  // Only when both Op0 and Op1 have uses, this fold produces one extra
+  // instruction.
+  if (match(Op0, m_Shl(m_Value(X), m_Value(Y))) &&
+      match(Op1, m_Shl(m_Specific(X), m_Value(Z))) &&
+      (Op0->hasOneUse() || Op1->hasOneUse())) {
```

Comment:
maybe @nikic cay more, but I'd be okay with dropping the use check to get rid of the `div`. We already increase inst count to drop `div` in other places.

Reply:
Yeah, I think that would be fine.